simtfl/bc/__init__.py

```diff
@@ -0,0 +1,115 @@
+        self.fee = fee
+        self.issuance = issuance
+
+    def input(self, index):
+        """Returns the input with the given index."""
+        return self.inputs[index]
+
+    def output(self, index):
+        """Returns the output with the given index."""
+        return self.outputs[index]
+
+    def is_coinbase(self):
+        """Returns `True` if this is a coinbase transaction (it has no inputs)."""
+        return len(self.inputs) == 0
+
+
+class BCContext:
+    def __init__(self):
+        """Constructs an empty `BCContext`."""
+        self.transactions = deque()
+        self.utxo_set = set()
+        self.total_issuance = 0
+
+    def is_valid(self, tx):
+        """Is `tx` valid in this context?"""
+        return set(tx.inputs).issubset(self.utxo_set)
+
+    def add_if_valid(self, tx):
+        """
+        If `tx` is valid in this context, add it to the context and return `True`.
+        Otherwise leave the context unchanged and return `False`.
+        """
+        txins = set(tx.inputs)
+        valid = txins.issubset(self.utxo_set)
+        if valid:
+            self.utxo_set -= txins
+            self.utxo_set |= set(tx.outputs)
+            self.total_issuance += tx.issuance
+            self.transactions.append(tx)
+
+        return valid
+
+
+class BCBlock:
+    def __init__(self, parent, added_score, transactions, allow_invalid=False):
+        """
+        Constructs a `BCBlock` with the given parent block, score relative to the parent, and transactions.
+        If `allow_invalid` is set, the block need not be valid.
+        Use `parent=None` to construct the genesis block.
+        """
+        self.parent = parent
+        self.score = (0 if parent is None else self.parent.score) + added_score
+        self.transactions = transactions
+        self.hash = Unique()
+        assert allow_invalid or self.is_valid()
+
+    def is_valid(self):
+        return (
+            len(self.transactions) > 0 and
+            self.transactions[0].is_coinbase() and
+            not any((tx.is_coinbase() for tx in self.transactions[1:])) and
+            sum((tx.fee for tx in self.transactions)) == 0
+        )
+
+
+@dataclass
+class BCProtocol:
+    Transaction: type[object] = BCTransaction
+    Context: type[object] = BCContext
+    Block: type[object] = BCBlock
+
+
+class TestBC(unittest.TestCase):
+    def test_basic(self):
+        ctx = BCContext()
+        issuance_tx0 = BCTransaction([], [10], 0, issuance=10)
+        assert ctx.add_if_valid(issuance_tx0)
+        genesis = BCBlock(None, 1, [issuance_tx0])
+        assert genesis.score == 1
+
+        issuance_tx1 = BCTransaction([], [6], -1, issuance=5)
+        spend_tx = BCTransaction([issuance_tx0.output(0)], [9], 1)
+        assert ctx.add_if_valid(issuance_tx1)
+        assert ctx.add_if_valid(spend_tx)
+        block1 = BCBlock(genesis, 1, [issuance_tx1, spend_tx])
+        assert block1.score == 2
```

Comment:
Do we need the simulation to distinguish between validity failure cases? If so, let's add test coverage that ensures these edge cases lead to invalid blocks:

- Blocks:
  - 0 txns
  - non-coinbase initial txn
  - multiple coinbase txns
  - unbalanced fees (note: Doesn't Zcash NU5 mainnet allow unbalanced fees where coinbase claims less than the total allowed? Probably safe to ignore unless there were any possible security impacts for that edge case.)
  - contains "context-free-invalid" txns
  - contains txns with nonexistent (e.g. out-of-`BCContext` inputs)
  - contains txns with multi-spends (e.g. >1 txn with the same input).

Reply:
Yes, I intend to improve the test coverage later.